rules/match.go

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package rules
+
+import (
+	"github.com/m3db/m3metrics/rules"
+	"github.com/m3db/m3x/errors"
+)
+
+// Matcher matches rules against metric IDs.
+type Matcher interface {
+	// Match matches rules against metric ID and returns the match result.
+	Match(id []byte) rules.MatchResult
+
+	// Close closes the matcher.
+	Close() error
+}
+
+type matcher struct {
+	opts        Options
+	namespaceFn NamespaceFn
+
+	namespaces  *namespaces
+	cache       Cache
+}
+
+// NewMatcher creates a new rule matcher.
+func NewMatcher(cache Cache, opts Options) (Matcher, error) {
+	key := opts.NamespacesKey()
+	namespaces := newNamespaces(key, cache, opts)
+	if err := namespaces.Start(); err != nil {
+		return nil, err
+	}
+
+	return &matcher{
+		opts:        opts,
+		namespaceFn: opts.NamespaceFn(),
+		namespaces:  namespaces,
+		cache:       cache,
+	}, nil
+}
+
+func (m *matcher) Match(id []byte) rules.MatchResult {
+	namespace := m.namespaceFn(id)
```

Comment:
As discussed we will need to change the interfaces here so that something like `ReportableMetric` can be queried for a tag, e.g. `TagValue(name string)`.  This way statsd and any othe rmetric being represented in the future can be matched into a namespace.